### DISCUSSION TITLE
camelcase <-> snakecase notation

### DISCUSSION THREAD

**Discussion by mits87** [prisma/prisma#2530] [Q&A] [answered]
Hi, 

I would like to know is it possible to write schema like that: 

```
model Post {
  id        Int     @id @default(autoincrement())
  isActive Boolean @default(false)
  author    User?   @relation(fields:  [authorId], references: [id])
  authorId  Int?
}
```

but my MySQL schema should looks like that:

```
CREATE TABLE posts (
    id INT NOT NULL AUTO_INCREMENT,
    is_active BOOL DEFAULT false,
    author_id INT(11),
   PRIMARY KEY ( id )
);
```

I mean is there any implementation of camelcase to snakecase notation?
How can I achieve that?

Currently this is a problem for me because in my nodejs application everywhere I would like use camelcase notation but my old database still have snakecase. 

Thanks in advance for the answer.

**Comment by ryands17** [comment, marked as answer]
Hey @mits87 :wave: 
You can use the `@map` and `@@map` directives to rename the fields and table respectively to suit camelCase.

For e.g.

```sql
create table posts (
	id serial primary key,
	is_active boolean default false,
	author_id int
);
```
add the above table in your DB and then run `prisma introspect`.

You will get something like this:

```prisma
model posts {
  id        Int      @default(autoincrement()) @id
  author_id Int?
  is_active Boolean? @default(false)
}
```

Now add the directives as follows:

```prisma
model Post {
  id       Int      @default(autoincrement()) @id
  authorId Int?     @map("author_id")
  isActive Boolean? @default(false) @map("is_active")

  @@map("posts")
}
```

This will keep the API in camelCase and the db in snake_case.

**Reply by Sytten** [reply]
I think a default would go a long way honestly, not sure if that was requested before. I am delaying this work myself to see if we get that default before I run out of time to do it by hand :D

**Reply by ryands17** [reply]
True that. Currently I am using a post introspect script that handles this. It's not ideal, but should do for the timebeing.

**Reply by Sytten** [reply]
Care to post it here?

**Reply by ryands17** [reply]
Something like this:

```ts
import * as fs from 'fs'
import * as path from 'path'
import { camelCase, upperFirst } from 'lodash'
import * as pluralize from 'pluralize'

const PRISMA_FILE_PATH = path.join(__dirname, 'schema.prisma')

function snakeToCamel(str: string) {
  return camelCase(str)
}
function snakeToPascal(str: string) {
  return upperFirst(camelCase(str))
}

const PRISMA_PRIMITIVES = ['String', 'Boolean', 'Int', 'Float', 'DateTime']
const KNEX_INTERNAL_MODELS = [
  'knex_migrations',
  'knex_migrations_lock',
  'pgmigrations',
]

function isKnexInternalModel(typeName: string) {
  return KNEX_INTERNAL_MODELS.includes(typeName)
}

function isPrimitiveType(typeName: string) {
  return PRISMA_PRIMITIVES.includes(typeName)
}

function fixFieldsArrayString(fields: string) {
  return fields
    .split(', ')
    .map((field) => snakeToCamel(field))
    .join(', ')
}

async function fixPrismaFile() {
  const text = await fs.promises.readFile(PRISMA_FILE_PATH, 'utf8')

  const textAsArray = text.split('\n')

  const fixedText = []
  let currentModelName: string | null = null
  let hasAddedModelMap = false

  for (let line of textAsArray) {
    // Are we at the start of a model definition
    const modelMatch = line.match(/^model (\w+) {$/)
    if (modelMatch) {
      currentModelName = modelMatch[1]
      if (isKnexInternalModel(currentModelName)) {
        continue
      }
      hasAddedModelMap = false
      const pascalModelName = snakeToPascal(currentModelName)
      fixedText.push(`model ${pascalModelName} {`)
      continue
    }

    if (currentModelName && isKnexInternalModel(currentModelName)) {
      continue
    }

    // We don't need to change anything if we aren't in a model body
    if (!currentModelName) {
      fixedText.push(line)
      continue
    }

    // Add the @@map to the table name for the model
    if (!hasAddedModelMap && (line.match(/\s+@@/) || line === '}')) {
      if (line === '}') {
        fixedText.push('')
      }
      fixedText.push(`  @@map("${currentModelName}")`)
      hasAddedModelMap = true
    }

    // Renames field and applies a @map to the field name if it is snake case
    // Adds an s to the field name if the type is an array relation
    const fieldMatch = line.match(/\s\s(\w+)\s+(\w+)(\[\])?/)
    let fixedLine = line
    if (fieldMatch) {
      const [, currentFieldName, currentFieldType, isArrayType] = fieldMatch

      let fixedFieldName = snakeToCamel(currentFieldName)
      if (isArrayType && !pluralize.isPlural(fixedFieldName)) {
        fixedFieldName = pluralize.plural(fixedFieldName)
      }

      fixedLine = fixedLine.replace(currentFieldName, fixedFieldName)

      // Add map if we needed to convert the field name and the field is not a relational type
      // If it's relational, the field type will be a non-primitive, hence the isPrimitiveType check
      if (currentFieldName.includes('_') && isPrimitiveType(currentFieldType)) {
        fixedLine = `${fixedLine} @map("${currentFieldName}")`
      }
    }

    // Capitalizes model names in field types
    const fieldTypeMatch = fixedLine.match(/\s\s\w+\s+(\w+)/)
    if (fieldTypeMatch) {
      const currentFieldType = fieldTypeMatch[1]
      const fieldTypeIndex = fieldTypeMatch[0].lastIndexOf(currentFieldType)
      const fixedFieldType = snakeToPascal(currentFieldType)
      const startOfLine = fixedLine.substr(0, fieldTypeIndex)
      const restOfLine = fixedLine.substr(
        fieldTypeIndex + currentFieldType.length
      )
      fixedLine = `${startOfLine}${fixedFieldType}${restOfLine}`
    }

    // Changes `fields: [relation_id]` in @relation to camel case
    const relationFieldsMatch = fixedLine.match(/fields:\s\[([\w,\s]+)\]/)
    if (relationFieldsMatch) {
      const fields = relationFieldsMatch[1]
      fixedLine = fixedLine.replace(fields, fixFieldsArrayString(fields))
    }

    // Changes fields listed in @@index or @@unique to camel case
    const indexUniqueFieldsMatch = fixedLine.match(/@@\w+\(\[([\w,\s]+)\]/)
    if (indexUniqueFieldsMatch) {
      const fields = indexUniqueFieldsMatch[1]
      fixedLine = fixedLine.replace(fields, fixFieldsArrayString(fields))
    }

    fixedText.push(fixedLine)
  }

  await fs.promises.writeFile(PRISMA_FILE_PATH, fixedText.join('\n'))
}

fixPrismaFile()
```

Do let me know if this can be enchanced in any way.

**Reply by nikolasburk** [reply]
Here's a related feature request: https://github.com/prisma/prisma/issues/2425

![image](https://user-images.githubusercontent.com/4058327/83115117-33bbb900-a0ca-11ea-8fc0-79bf86ae7198.png)

**Reply by Inayat-C** [reply]
Just putting this here if anyone wants to quickly replace these import statements with require

> ```
> import * as fs from 'fs'
> import * as path from 'path'
> import { camelCase, upperFirst } from 'lodash'
> import * as pluralize from 'pluralize'
> 
> ```

```
const fs = require('fs');
const path = require('path');
const pluralize = require('pluralize');
const camelCase = require('lodash.camelcase');
const upperFirst = require('lodash.upperfirst');
```

**Reply by lvjiaxuan** [reply]
up-to-date

```js
const fs = require('fs');
const path = require('path');
const pluralize = require('pluralize');
const camelCase = require('lodash/camelcase');
const upperFirst = require('lodash/upperfirst');
```

**Comment by lvjiaxuan** [comment]
https://paljs.com/cli/schema/

**Comment by tofsjonas** [comment]
https://github.com/iiian/prisma-case-format

**Reply by olivierlacan** [reply]
@tofsjonas This is exactly what I came here to ask, thank you! 

It would fantastic if we could configure the introspection formatting at the schema level to save a *ton* of time manually mapping all the fields in our app from snake_case (db convention) to camelCase (app convention). This tool does what I expected would be available as a Prisma configuration like this: 

```typescript
datasource db {
  provider     = "postgresql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
  extensions   = [uuid_ossp(map: "uuid-ossp", schema: "public")]
  schemas      = ["ext", "public"]
  fieldNaming  = { "db": "snake_case", "schema": "camelCase" } 
}
```

Where `"db"` defines the convention used in the database schema itself for field naming, and where the `"schema"` value would result in `@map` statements being added for all fields that match the `db` format.

Example: 

```typescript
createdAt      DateTime         @default(now()) @db.Timestamptz(6) @map("created_at")
```


Granted I know that @janpio mentioned y'all are trying to not add more configuration options at this time, but since Prisma is already a fantastic time-saver on so many fronts, this feels like a quality of life improvement worth chasing. 😅

**Reply by dchenk** [reply]
@olivierlacan's idea would be great! Another way to possibly configure this could be with a `defaultMapFormat` option in the `schema.prisma` file, and if any particular field or column needs an override it could have its own `@map` or `@@map` directive.

**Reply by jiahui92** [reply]
great, it work well to me. i found the other 2 tools and not work well enought for what i need

**Comment by ludralph** [comment]
Hi @mits87,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by allenkaplan** [comment]
Relevant prisma docs https://www.prisma.io/docs/orm/prisma-schema/data-model/database-mapping
